Log documentation (#310)
avenga/couper: docs/LOGS.md

- [Introduction](#introduction)
  - [Log Types](#log-types)
  - [Fields](#fields)
    - [Access Fields](#access-fields)
    - [Backend Fields](#backend-fields)
    - [Daemon Fields](#daemon-fields)
  - [Features](#features)

# Introduction

> _Logs_ are the information being printed in the console upon the execution of Couper. There are a few different [Log Types](#log-types) and several different [Fields](#fields) containing useful information given in said _Logs_ to be remembered, but you can adjust the _Logs'_ appearance and verbosity with a variety of [Features](#features) to make sure that the information you seek is available as easily as possible.

Reviewer: ```suggestion
_Logs_ are the information being printed in the console upon the execution of Couper. There are a few different [Log Types](#log-types) and several different [Fields](#fields) containing useful information given in said _Logs_ to be remembered, but you can adjust the _Logs'_ appearance and verbosity with a variety of [Settings](#Settings) to make sure that the information you seek is available as easily as possible.
```
lets call it settings since its related to the configuration settings block.

These _Logs_ are however still subject to change. We aspire to make Couper as stable as possible and that includes the _Logs_, so until we believe to have achieved that goal you should expect to see some changes.

## Log Types

| Type                | Description                                                                                                                                     |
| :------------------ | :---------------------------------------------------------------------------------------------------------------------------------------------- |

> | `couper_access`     | Provides information about the frontend side of things. For information about its [Fields](#fields), see [Access Fields](#access-fields).       |

Reviewer: Some additional information: All client requests and final response status-code?

> | `couper_access_tls` | Provides information about connection to TLS proxies. For information about its [Fields](#fields), see [Access Fields](#access-fields). |

Reviewer: ```suggestion
| `couper_access_tls` | Provides information about connections via configured [https_dev_proxy](./REFERENCE.md#settings). For information about its [Fields](#fields), see [Access Fields](#access-fields). |
```
just client requests which requesting via configured tls dev proxy port.

> | `couper_backend`    | Provides information about the backend side of things. For information about its [Fields](#fields), see [Backend Fields](#backend-fields).      |

Reviewer: Add some additional information like upstream request / response states. Origin informations are useful for example.

| `couper_daemon`     | Provides background information about the execution of Couper. It is here where the main purpose of the `message` [Field](#fields) lies, as each printed _Log_ of this type will contain a `message` entry containing a description of the current actions of Couper. For information about its [Fields](#fields), see [Daemon Fields](#daemon-fields).                                                                                                                                                       |

## Fields

> Given the large amount of information contained in several _Logs_, see [Features](#features) to change the format or level of intricacy with which _Logs_ are printed.

Reviewer: Same here, settings.

### Common Fields

These are a part of all [Log Types](#log-types) and are hence mentioned seperately to avoid unnecessary redundancy.

| Name          | Description                                       |
| :------------ | :------------------------------------------------ |

> | `"build”`     | git short hash                                    |

Reviewer: ```suggestion
| `"build”`     | git short hash during build time                                    |
```

| `"response": {`           | field regarding response information                                                                                         |
| &nbsp;&nbsp;`"bytes"`     | body size of response                                                                                                        |
| &nbsp;&nbsp;`"headers" }` | field regarding keys and values originating from configured keys/header names                                                |
| `"server"`                | server name (defined in couper file)                                                                                         |
| `"status"`                | response status code                                                                                                         |
| `"timings": {`            | field regarding timing                                                                                                       |
| &nbsp;&nbsp;`"total" }`   | total time taken                                                                                                             |

> | `“uid"`                   | unique request id configurable in [Settings](../docs/REFERENCE.md#settings-block)                                            |

Reviewer: Applies to all `../docs/*` appearances: I think `./*` would be enough since this document lies next to the referenced ones.

| `"timings": {`           | field regarding timing                                                                                                       |
| &nbsp;&nbsp;`"dns"`      | time taken by dns                                                                                                            |
| &nbsp;&nbsp;`"tcp"`      | time taken between attempting and establishing tcp connection                                                                |
| &nbsp;&nbsp;`"tls"`      | time taken between attempt and success at tls handshake                                                                      |
| &nbsp;&nbsp;`"total"`    | total time taken                                                                                                             | 
| &nbsp;&nbsp;`"ttfb" }`   | time to first byte/between establishing connection and receiving first byte                                                  |
| `"token_request"`        | entry regarding request for token                                                                                            |

> | `"token_request_retry"`  | how many attempts at `token_request`                                                                                         |

Reviewer: ```suggestion
| `"token_request_retry"`  | how many `token_request` attempts were made                                                                                   |
```

| Name                          | Description                                                                                                                |
| :---------------------------- | :------------------------------------------------------------------------------------------------------------------------- |
| `"deadline"`                  | shutdown parameter, see [Health-Check](./REFERENCE.md#health-check)                                                        |
| `"delay"`                     | shutdown parameter, see [Health-Check](./REFERENCE.md#health-check)                                                        |
| `"watch": {`                  | field watching configuration file changes, logs with this field only appear if `watch=true`, more in [Settings](#settings) |
| &nbsp;&nbsp;`"max-retries",`  | maximum retry count, see [Global Options](./CLI.md#global-options)                                                         |
| &nbsp;&nbsp;`"retry-delay" }` | configured delay of each retry, see [Global Options](./CLI.md#global-options)                                              |

## Settings

> For more information regarding the usage of these settings, see the documentations regarding [Command Line Interface](./CLI.md#global-options), [Environment Options](./../DOCKER.md#environment-options) and/or [All Settings](./REFERENCE.md#settings-block).

Reviewer 2: ```suggestion
For more information regarding the usage of these settings, compare the [Command Line Interface](./CLI.md#global-options) documentation, Couper's [Environment Options](./../DOCKER.md#environment-options) and/or the [Settings Block](./REFERENCE.md#settings-block) reference.
```
